WeakMap and WeakSet (#184)
javascript-tutorial/it.javascript.info: 1-js/05-data-types/08-weakmap-weakset/article.md

# WeakMap e WeakSet

> Come abbiamo già compreso nel capitolo <info:garbage-collection>, il motore JavaScript mantiene un valore in memoria fino a che questo risulta accessibile (e potrebbe potenzialmente essere utilizzato).

Reviewer: ```suggestion
Come abbiamo già visto nel capitolo <info:garbage-collection>, il motore JavaScript mantiene un valore in memoria fino a che questo risulta accessibile (e potrebbe potenzialmente essere utilizzato).
```
In questo contesto 'compreso' non suona tanto bene.

Ad esempio:
```js
let john = { name: "John" };

// l'oggetto è accessibilie, john è un suo riferimento

// sovrascriviamo il riferimento
john = null;

*!*
// l'oggetto verrà rimosso dalla memoria
*/!*
```

> Solitamente, le proprietà di un oggetto o gli elementi di un array o di qualsiasi altra struttura dati vengono considerati accessibili e mantenuti in memoria fino a che la struttura di dati rimane in memoria.

Reviewer: ```suggestion
Solitamente, le proprietà di un oggetto o gli elementi di un *array* o di qualsiasi altra *data structure* vengono considerati accessibili e mantenuti in memoria fino a che la *data structure* rimane in memoria.
```

Io qui metterei il termine tecnico, 'struttura dei dati' puo' facilmente confondere e poi *data structure* e' un termine tecnico che il lettore dovrebbe comprendere il prima possibile. Che ne pensi?

> Solitamente, le proprietà di un oggetto o gli elementi di un array o di qualsiasi altra struttura dati vengono considerati accessibili e mantenuti in memoria fino a che la struttura di dati rimane in memoria.

Reviewer: ```suggestion
Solitamente, le proprietà di un oggetto o gli elementi di un array o di qualsiasi altra struttura dati vengono considerati accessibili e mantenuti in memoria fino a che la struttura dati rimane in memoria.
```

> Ad esempio, se inseriamo un oggetto in un array, fino a che l'array rimane vivo, anche l'oggetto rimarrà vivo, anche se non sono presenti riferimenti.

Reviewer: ```suggestion
Ad esempio, se inseriamo un oggetto in un array, fino a che l'array rimane "vivo", anche l'oggetto rimarrà in memoria, anche se non sono presenti riferimenti.
```
Siamo un po' piu' pignoli degli autori della versione inglese :)

Come nell'esempio:

```js
let john = { name: "John" };

let array = [ john ];

john = null; // sovrascriviamo il riferimento

*!*
// john è memorizzato all'interno dell'array
// quindi non verrà toccato dal garbage collector

> // possiamo ottenerlo tramite array[0]

Reviewer 2: ```suggestion
// possiamo estrarlo tramite array[0]
```

*/!*
```

O, se utilizziamo un oggetto come chiave in un `Map`, fino a che la `Map` esiste, anche l'oggetto esisterà. Occuperà memoria e non potrà essere ripulito dal garbage collector.

Ad esempio:

```js
let john = { name: "John" };

let map = new Map();
map.set(john, "...");

john = null; // sovrascriviamo il riferimento

*!*
// john viene memorizzato all'interno di map,

> // possiamo ottenerlo utilizzando map.keys()

Reviewer 2: ```suggestion
// possiamo estrarlo utilizzando map.keys()
```

Reviewer 2: come sopra, ho usato "estrarre" ... non so se tecnicamente sia appropriato, valuta tu

*/!*
```

> `WeakMap` è fondamentalmente diverso in questo aspetto. Infatti non previene l'azione del garbage collector sugli oggetti utilizzati come chiave.

Reviewer: ```suggestion
`WeakMap` è fondamentalmente diverso in questo aspetto. Infatti non previene l'azione del garbage collector sugli oggetti utilizzati come chiave.
```
```suggestion
`WeakMap` è fondamentalmente diverso sotto questo aspetto. Infatti non previene l'azione del garbage collector sugli oggetti utilizzati come chiave.
```

Vediamo cosa significa questo, utilizzando degli esempi.

## WeakMap

La prima differenza tra `Map` e `WeakMap` è che le chiavi devono essere oggetti, non valori primitivi:

```js run
let weakMap = new WeakMap();

let obj = {};

weakMap.set(obj, "ok"); // funziona

*!*
// non possiamo utilizzare una stringa come chiave
weakMap.set("test", "Whoops"); // Errore, perché "test" non è un oggetto
*/!*
```

> Ora, se utilizziamo un oggetto come chiave, e perdiamo tutti i riferimenti a quell'oggetto -- questo verrà rimosso dalla memoria (e dalla map) automaticamente.

Reviewer 2: ```suggestion
Ora, se utilizziamo un oggetto come chiave, e non ci sono altri riferimenti a quell'oggetto -- questo verrà rimosso dalla memoria (e dalla map) automaticamente.
```

```js
let john = { name: "John" };

let weakMap = new WeakMap();
weakMap.set(john, "...");

john = null; // sovrascriviamo il riferimento

// john è stato rimossa dalla memoria!
```

Confrontiamolo con l'esempio di `Map` visto sopra. Ora se `john` esiste solo come chiave della `WeakMap` -- verrà eliminato automaticamente dalla map (e anche dalla memoria).

> `WeakMap` non supportano gli iteratori e i metodi `keys()`, `values()`, `entries()`, quindi non c'è alcun modo di ottenere tutte le chiavi o valori tramite questi metodi.

Reviewer 2: ```suggestion
`WeakMap` non supporta gli iteratori e i metodi `keys()`, `values()`, `entries()`, quindi non c'è alcun modo di ottenere tutte le chiavi o valori tramite questi metodi.
```

Dove potremmo avere bisogno di una struttura simile?

## Caso d'uso: dati aggiuntivi

Il principale campo di applicazione di `WeakMap` è quello di un *additional data storage*.

Se stiamo lavorando con un oggetto che "appartiene" ad un altro codice, magari una libreria di terze parti, e vogliamo memorizzare alcuni dati associati ad esso, che però dovrebbero esistere solamente finché l'oggetto esiste - allora una `WeakMap` è proprio ciò di cui abbiamo bisogno.

Inseriamo i dati in una `WeakMap`, utilizzando l'oggetto come chiave, quando l'oggetto verrà ripulito dal garbage collecetor, anche i dati associati verranno ripuliti.

```js
weakMap.set(john, "secret documents");
// se john muore, i documenti segreti verranno distrutti automaticamente
```

Proviamo a guardare un esempio.

> Possiamo pensare di avere del codice che tiene nota del conteggio delle visite per ogni utente. L'informazioni viene memorizzata in un map: l'utente è la chiave mentre il conteggio delle visite è il valore. Quando l'utente esce, vogliamo smettere di mantenere in memoria il conteggio delle visite.

Reviewer 2: ```suggestion
Immaginiamo di avere del codice che tiene nota del numero di visite per ogni utente. L'informazioni viene memorizzata in un map: l'utente è la chiave, mentre il conteggio delle visite è il valore. Quando l'utente esce, vogliamo smettere di mantenere in memoria il conteggio delle visite.
```

let pete = { name: "Pete" };
let mary = { name: "Mary" };

visitedSet.add(john); // John ci ha visitato
visitedSet.add(pete); // Poi Pete
visitedSet.add(john); // John di nuovo

// visitedSet ha 2 utenti ora

// controlliamo se John ci ha visitato
alert(visitedSet.has(john)); // true

// controlliamo se Mary ci ha visitato
alert(visitedSet.has(mary)); // false

john = null;

// visitedSet verrà ripulito automaticamente
```

La maggior limitazione di `WeakMap` e `WeakSet` è l'assenza di iteratori, e la mancanza della possibilità di ottenere tutti gli elementi contenuti. Potrebbe sembrare un inconveniente, ma non vieta a `WeakMap/WeakSet` di compiere il proprio lavoro -- essere una struttura "addizionale" per memorizzare informazioni relative a dati memorizzati in un altro posto.

## Riepilogo

`WeakMap` è una collezione simile a `Map`, che permette di utilizzare solamente gli oggetti come chiave, con la differenza che la rimozione di un oggetto, rimuove anche il valore associato.

`WeakSet` è una collezione simile a `Set`, che memorizza solamente oggetti, e li rimuove completamente una volta che diventano inaccessibili.

> I loro principali vantaggi sono che possiedono un riferimento debole agli oggetti, in questo modo possono essere facilmente ripuliti dal garbage collector.

Reviewer 2: ```suggestion
Il loro principale vantaggio è che possiedono un riferimento debole agli oggetti, in questo modo possono essere facilmente ripuliti dal garbage collector.
```
qui è usato il plurale, anche in inglese, ma leggo un solo "vantaggio"

> Questi vantaggi però non consento l'utilizzo di `clear`, `size`, `keys`, `values`...

Reviewer 2: ```suggestion
Il lato negativo è di non poter utilizzare `clear`, `size`, `keys`, `values`...
```

> `WeakMap` e `WeakSet` vengono utilizzate come strutture dati "secondarie" in aggiunta a quelle "principali". Una volta che l'oggetto viene rimossa dalla struttura dati "principale", se l'unico riferimento rimasto è una chiave di `WeakMap` o `WeakSet`, allora verranno rimossi.

Reviewer 2: ```suggestion
`WeakMap` e `WeakSet` vengono utilizzati come strutture dati "secondarie" in aggiunta a quelle "principali". Una volta che l'oggetto viene rimosso dalla struttura dati "principale", se l'unico riferimento rimasto è una chiave di `WeakMap` o `WeakSet`, allora verrà rimosso.
```
